### PR TITLE
refactor ./securedrop-admin sdconfig from Ansible to python. 

### LINKED ISSUE #2522: refactor securedrop-admin sdconfig in python
# Feature request

## Description

securedrop-admin sdconfig is implemented in python via install_files/ansible-base/securedrop-configure.yml

ansible is contorted and not user friendly in this context. What we want is to display a prompt to the user with a default or the previous value read from site_specific. And have immediate validation instead of post-submission validation.

The [click](http://click.pocoo.org/5/) library is already a SecureDrop dependency and could be used for that purpose.

See also https://github.com/freedomofpress/securedrop/issues/2266 and https://github.com/freedomofpress/securedrop/issues/2095

## User Stories

As an admin I want to get immediate validation when `securedrop-admin sdconfig` prompts me for a valid when configuring securedrop for the first time.

As an admin I want `securedrop-admin sdconfig` to display the values from `install_files/ansible-base/group_vars/all/site-specific` when running it the second time so that I can modify them instead of editing the file manually.

### PR DESCRIPTION
**Pull request by dachary**
## Status

Ready for Review

## Description of Changes

Fixes #2522

Refactor ``./securedrop-admin sdconfig`` from Ansible to python. 

### Act I

The commits up to **securedrop-admin: refactor sdconfig as a python script** included implement the basic features **securedrop-admin** needs to replace the YAML implementation. It produces the ``site-specific`` file will all know and love. With the nice addition that it can also be used to update it. The current value is displayed for the user to edit (with emacs & vi bindings) and is validated as soon as the user hits enter. Verifications involving multiple variables (i.e. fingerprint and public key) are done at the very end.

### Act II

Each **securedrop-admin: refactor sdconfig ABC** commit is a replacement for the implementation of the corresponding YAML variable found in ``site-specific``. Validation classes are implemented when a variable needs it and not in a different commit although they can be re-used later. Each variable is implemented in the order in which they originally showed in the Ansible implemenation. With the notable exception of the **securedrop-admin: define ValidatePath** commit which is a reminder of the ancient time when the logo was updated from **securedrop-admin** instead of the journalist admin interface.

### Act III

Over time **securedrop-admin** grew from a tiny standalone script isolated at the root of the git repository to a handsome young module who moved with its relatives down in the securedrop directory. Alas, its users were different and it became clear **securedrop-admin** had to move to its own directory: **admin**. From **ansible: trim validate role comment now obsolete** up it got a setup.py, requirements and tox to run the tests, a Dockerfile and a Makefile to help the CI. It also was split in two: the **setup** action is now a standlone **bootstrap.py** script instead of being included in a rather contorted way into a single standalone script.

### Epilogue

One day **securedrop-admin** could be uploaded to pypi and have its own repository even. But that's a story for another time. 

## Testing

In both Tails and the development environment, try to enter invalid (non existent domain name) or inconsistent (fingerprint not matching the public key) values to check the resulting  interaction.

### In the Debian GNU/Linux stretch development environment

* ./securedrop-admin setup
* ./securedrop-admin sdconfig
* fill all values
* verify `install_files/ansible-base/group_vars/all/site-specific` has the desired variables and values

### In Tails
* setup tails with persistence
* git clone -b wip-dachary-2522-sdconfig https://github.com/freedomofpress/securedrop
* cd securedrop
* ./securedrop-admin setup
* ./securedrop-admin sdconfig
* verify `install_files/ansible-base/group_vars/all/site-specific` has the desired variables and values
* ./securedrop-admin sdconfig # and change one value
* vierify `install_files/ansible-base/group_vars/all/site-specific` has the desired variables and values

### In CI

* verify the new .circleci/config.yml target admin-tests actually runs

## Deployment

It is supposed to be a backward compatible replacement of `./securedrop-admin sdconfig``, reason why the documentation is not modified.

### If you made changes to the app code:

- [x] Unit and functional tests pass on the development VM

### If you made changes to documentation:

- [x] Doc linting passed locally

### CODE REVIEW

**Comment by dachary** [comment]
@msheiny it's mostly done, modulo tiny details and ready for a review :-) You would want to start from **
securedrop-admin: remove comment with no useful information** because the commits before that are from another PR which is not merged yet. Enjoy !

**Comment by dachary** [comment]
transient error https://travis-ci.org/freedomofpress/securedrop/builds/320447653?utm_source=github_status&utm_medium=notification

<pre>
xs-perl (2.340-1build1) ...", "Setting up liblist-moreutils-perl (0.33-1build3) ...", "Setting up libnet-http-perl (6.06-1) ...", "Setting up libwww-robotrules-perl (6.01-1) ...", "Setting up libperlio-gzip-perl (0.18-1build3) ...", "Setting up libtext-levenshtein-perl (0.06~01-2) ...", "Setting up libtie-ixhash-perl (1.23-1) ...", "Setting up hardening-includes (2.5ubuntu2.close failed in file object destructor:
sys.excepthook is missing
lost sys.stderr
The command "ansible-playbook -i inventory -vv --connection=local install_files/ansible-base/securedrop-development.yml
  " failed and exited with 1 during .
Your build has been stopped.
</pre>

**Comment by dachary** [comment]
transient error https://circleci.com/gh/freedomofpress/securedrop/5457?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

<pre>
Fetched 60.2 MB in 18s (3300 kB/s)
E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/c/curl/libcurl3-gnutls_7.35.0-1ubuntu2.12_amd64.deb  404  Not Found [IP: 91.189.91.23 80]
</pre>

**Comment by dachary** [comment]
Seen that CircleCI error often in the past two days (twice on this PR today) https://circleci.com/gh/freedomofpress/securedrop/5851?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link 

<pre>
Get:157 http://archive.ubuntu.com/ubuntu/ trusty-updates/main gnupg2 amd64 2.0.22-3ubuntu1.3 [681 kB]
Fetched 60.2 MB in 22s (2673 kB/s)
E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/c/curl/libcurl3-gnutls_7.35.0-1ubuntu2.12_amd64.deb  404  Not Found [IP: 91.189.88.152 80]

E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/libx/libxml2/libxml2_2.9.1+dfsg1-3ubuntu4.10_amd64.deb  404  Not Found [IP: 91.189.88.152 80]
</pre>

**Comment by dachary** [comment]
Again CircleCI https://circleci.com/gh/freedomofpress/securedrop/5858?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

<pre>
Get:156 http://archive.ubuntu.com/ubuntu/ trusty/universe secure-delete amd64 3.1-6 [67.3 kB]
Get:157 http://archive.ubuntu.com/ubuntu/ trusty-updates/main gnupg2 amd64 2.0.22-3ubuntu1.3 [681 kB]
Fetched 60.2 MB in 18s (3305 kB/s)
E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/c/curl/libcurl3-gnutls_7.35.0-1ubuntu2.12_amd64.deb  404  Not Found [IP: 91.189.88.152 80]

E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/libx/libxml2/libxml2_2.9.1+dfsg1-3ubuntu4.10_amd64.deb  404  Not Found [IP: 91.189.88.152 80]

E: Failed to fetch http://security.ubuntu.com/ubuntu/pool/main/p/python2.7/libpython2.7-minimal_2.7.6-8ubuntu0.3_amd64.deb  404  Not Found [IP: 91.189.88.161 80]
</pre>

**Comment by dachary** [comment]
travis passed after only 25 restart :-)

**Comment by dachary** [comment]
rebased (thanks for noticing @heartsucker) and resolved conflicts due to a change in the validation process of sdconfig in the YAML implementation

**Comment by dachary** [comment]
@msheiny in case that can be useful to understand what's going on, here is another instance https://circleci.com/gh/freedomofpress/securedrop/6340?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

<pre>
    TASK [tor-hidden-services : Setup Tor apt repo.] *******************************
    Thursday 11 January 2018  20:47:08 +0000 (0:00:01.075)       0:03:53.155 ******
    fatal: [mon-staging]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Shared connection to 13.56.58.218 closed.\r\n", "module_stdout": "Traceback (most recent call last):\r\n  File \"/tmp/ansible_ELhdhw/ansible_module_apt_repository.py\", line 565, in <module>\r\n    main()\r\n  File \"/tmp/ansible_ELhdhw/ansible_module_apt_repository.py\", line 553, in main\r\n    cache.update()\r\n  File \"/usr/lib/python2.7/dist-packages/apt/cache.py\", line 440, in update\r\n    raise FetchFailedException(e)\r\napt.cache.FetchFailedException: W:Failed to fetch https://tor-apt.freedom.press/dists/trusty/main/binary-amd64/Packages  HttpError403\r\n, E:Some index files failed to download. They have been ignored, or old ones used instead.\r\n", "msg": "MODULE FAILURE", "rc": 0}
    changed: [app-staging]
</pre>

**Comment by dachary** [comment]
@redshiftzero would you have time to review this series ? It would be really nice to get it merged during this cycle and as early as possible to give us plenty of time to run into corner cases.

**Comment by dachary** [comment]
rebased & repushed with trivial conflict on config.yml

File: .circleci/config.yml [hunk 1]
```diff
@@ -77,6 +77,15 @@ jobs:
       - store_artifacts:
           path: ~/test-results
 
+  admin-tests:
+    docker:
+      - image: gcr.io/cloud-builders/docker
+    steps:
+      - run: apt-get install -y make
```
[the remark this answers is not in the document]
Comment (dachary) [reply .circleci/config.yml]: the tox command runs flake8, pylint and tests

File: admin/securedrop_admin/__init__.py [hunk 2]
```diff
@@ -0,0 +1,529 @@
...
+            if self.is_tails():
+                cmd = 'torify ' + cmd
+            cmd += ' ' + (dns and dns or '8.8.8.8')
+            try:
+                result = subprocess.check_output(cmd, shell=True,
+                                                 stderr=subprocess.STDOUT)
+            except subprocess.CalledProcessError as e:
+                result = e.output
+            sdlog.debug(cmd + ' => ' + result)
+            return 'has address' in result
```
[the remark this answers is not in the document]
Comment (dachary) [reply admin/securedrop_admin/__init__.py]: Ok, I'm removing these checks

File: securedrop/requirements/securedrop-app-code-requirements.in [hunk 3]
```diff
@@ -6,6 +6,7 @@ Flask
 gnupg
 Jinja2
 jsmin
+prompt_toolkit
```
[the remark this answers is not in the document]
Comment (dachary) [reply securedrop/requirements/securedrop-app-code-requirements.in]: Oh, good catch ! Unintended leftover, fixed in the **securedrop-admin: move to the standalone admin directory** commit where it belongs.

File: securedrop/requirements/admin-requirements.txt [hunk 4]
```diff
@@ -93,6 +93,9 @@ paramiko==2.1.2 \
     --hash=sha256:5fae49bed35e2e3d45c4f7b0db2d38b9ca626312d91119b3991d0ecf8125e310 \
     --hash=sha256:bdf239647e18b9b9ddbc2894fd1de9786b7a9144b1d19e32a5be3bb4bb63ae5d \
     # via ansible
+prompt-toolkit==1.0.15 \
```
[the remark this answers is not in the document]
Comment (dachary) [reply securedrop/requirements/admin-requirements.txt]: admin-requirements.txt is a unintended leftover, fixed in the securedrop-admin: move to the standalone admin directory commit where it belongs.

**Comment by dachary** [comment]
@redshiftzero addressed your comments, hungry for more :-)

**Comment by kushaldas** [comment]
I will look into this today.

**Comment by dachary** [comment]
@redshiftzero thanks for this review. It made me realize I underestimated border cases that show up frequently in real life (i.e. not in tests ;-). I did not modify the existing commits (except for the misplaced requirement inclusions which were just wrong) but worked from commit **
securedrop-admin: implement ValidateNotEmpty** to 

* address your comments
* improve verifications (required variables)
* catch keyboard interrupt
* display a message instead of nothing when an exception happens

This will hopefully be less broken now.

**Comment by dachary** [comment]
The file **securedrop_admin/__init__.py** could be renamed **securedrop_admin/main.py** or something if you prefer. Let me know which name suits you and I'll do it.

**Comment by redshiftzero** [comment]
Great, I've verified that the issues I found yesterday have been resolved (the filename is not an issue, I was just curious 😇 ). I think that this large diff would benefit from more eyes, so I'm going to wait until at least @kushaldas or @emkll has had an opportunity to take a look at this tomorrow prior to approving and merging in (if no other issues come up or no one else has the time to review, then I will stamp so we can get this into `develop` early in the release cycle).

**Comment by dachary** [comment]
> (the filename is not an issue, I was just curious

I could not find a name. And it did not seem right to split the large script into multiple files, one for each sub-command for instance (that would make sense). So, instead of giving it a meaningless name I opted to just have it in \_\_init__.py which is the equivalent of not giving it a name :-)

**Comment by kushaldas** [comment]
What am I missing here? I did `./securedrop-admin setup` before this. This is a normal debian box.

```
$ ./securedrop-admin sdconfig
./securedrop-admin: line 9: d: command not found
ERROR (run with -v for more): [Errno 2] No such file or directory: '/securedrop/translations'
```

**Comment by dachary** [comment]
> What am I missing here

My bad, I did cleanup this morning and introduced this regression. Please try again !

**Comment by dachary** [comment]
> If user provides the gmail SMTP address, then we should provide some default value for the port (as we already know that value).

This would require a significant additional work because there currently is no logic to change the default value of one field based on the value of another field. I thought about doing that for validation too but did not because that would add even more to this already large pull request. Do you think this is a blocker ?

**Comment by dachary** [comment]
@emkll good catch ! I pushed an additional commit to address this case (**securedrop-admin: display helpful message when no argument**). This discarded your approval and you'll need to restamp the PR if you agree with this modification :-)

<pre>
$ ./securedrop-admin 
please run './securedrop-admin setup' first
</pre>

**Comment by dachary** [comment]
Transient CircleCI error https://circleci.com/gh/freedomofpress/securedrop/7254?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

<pre>
Allocating a remote Docker Engine
Got error while creating host: rpc error: code = Unknown desc = An internal error occured when provisioning resources for this job.  Provisioning Service returned status code: 500
We had an unexpected error preparing a VM for this build, potentially due to our infrastructure or cloud provider.  Please retry the build in a few minutes
</pre>

**Comment by emkll** [comment]
@dachary : can you confirm that install, backup, logs and restore are working for you locally? When I try to run them on the current branch, I get the following:
```
amnesia@amnesia:~/Persistent/securedrop$ ./securedrop-admin -v install
INFO: Now installing SecureDrop on remote servers.
INFO: You will be prompted for the sudo password on the servers.
INFO: The sudo password is only necessary during initial installation.
Traceback (most recent call last):
  File "/home/amnesia/Persistent/securedrop/admin/.venv/bin/securedrop-admin", line 6, in <module>
    exec(compile(open(__file__).read(), __file__, 'exec'))
  File "/home/amnesia/Persistent/securedrop/admin/bin/securedrop-admin", line 22, in <module>
    main(sys.argv[1:])
  File "/home/amnesia/Persistent/securedrop/admin/securedrop_admin/__init__.py", line 539, in main
    args.func(args)
  File "/home/amnesia/Persistent/securedrop/admin/securedrop_admin/__init__.py", line 413, in install_securedrop
    '--ask-become-pass'], cwd=args.ansible_path)
  File "/usr/lib/python2.7/subprocess.py", line 181, in check_call
    retcode = call(*popenargs, **kwargs)
  File "/usr/lib/python2.7/subprocess.py", line 168, in call
    return Popen(*popenargs, **kwargs).wait()
  File "/usr/lib/python2.7/subprocess.py", line 390, in __init__
    errread, errwrite)
  File "/usr/lib/python2.7/subprocess.py", line 1024, in _execute_child
    raise child_exception
OSError: [Errno 2] No such file or directory
```

**Comment by dachary** [comment]
@emkll I'm so glad you caught this one ! When the commit **securedrop-admin: replace global variables by arguments** replaced all global variables by arguments, it did not convert them into absolute paths. Reason why you had this error. I amended this commit to do the right thing (i.e. os.path.realpath on each argument so they are all absolute paths, just like the global variables were before them).

I resisted the urge to write tests for the entire **securedrop-admin** script but I should have checked this manually, my bad.

**Comment by emkll** [comment]
@dachary it appears to be still throwing the same error for `./securedrop-admin logs`

**Comment by dachary** [comment]
> it appears to be still throwing the same error for ./securedrop-admin logs

Could you please show the full log of the shell session ? I can't reproduce the problem here.

**Comment by dachary** [comment]
@emkll another stupid mistake on my part. I verified on my laptop and not on tail and ansible is installed... but it is not in tails. Fixing the dependencies.

**Comment by dachary** [comment]
@emkll the problem this time was even bigger: ansible missing from the dependencies. I'll manually test in tails since we don't have (yet ;-) tests for securedrop-admin that verify ansible is available.

**Comment by dachary** [comment]
@emkll I verified manually that each subcommand successfully starts its work. I believe the untested part of securedrop-admin is ok because the only change that was made was to replace global variables with arguments. That can be checked visually in the **securedrop-admin: replace global variables by arguments** commit. The missing dependency to ansible was added by amending the **securedrop-admin: add requirements** and pinned to 2.3.2.

**Comment by dachary** [comment]
@emkll https://github.com/freedomofpress/securedrop/pull/2929 was merged and creates conflicts since we now depend on ansible 2.3.2. The **securedrop-admin: add requirements** commit was updated accordingly and nothing else was changed.

**Comment by dachary** [comment]
rebased (trivial conflict due to flake8 change in Makefile) and repushed !

**Comment by dachary** [comment]
@redshiftzero thanks ! And I stand ready to properly fix whatever problem is discovered during Q/A.

**Pull request merged**